## Jacknife

In [1]:
import numpy as np

In [2]:
def Jacknife(X , estimador, q=0.75):

#######################################################

    if estimador == 'median':

        def Jacknife_sample(X , r):

            X_sample_r = np.delete(X, r)

            return(X_sample_r)

        
        replicas_estimador = []

        for r in range(0, len(X)):

            replicas_estimador.append( np.median( Jacknife_sample(X , r) ) )

    
        n = len(X)

        sesgo = (n-1) * ( np.mean( replicas_estimador ) - np.median(X) )

        estimacion_sesgo_corregido = np.median(X) - sesgo

        standard_error = np.sqrt( ((n-1)/n) * sum( (replicas_estimador - np.mean( replicas_estimador ))**2 ) )

 #######################################################   

    if estimador == 'mean':

        def Jacknife_sample(X , r):

            X_sample_r = np.delete(X, r)

            return(X_sample_r)

        
        replicas_estimador = []

        for r in range(0, len(X)):

            replicas_estimador.append( np.mean( Jacknife_sample(X , r) ) )

    
        n = len(X)

        sesgo = (n-1) * ( np.mean( replicas_estimador ) - np.mean(X) )

        estimacion_sesgo_corregido = np.mean(X) - sesgo

        standard_error = np.sqrt( ((n-1)/n) * sum( (replicas_estimador - np.mean( replicas_estimador ))**2 ) )


#######################################################

    if estimador == 'std':

        def Jacknife_sample(Xj , r):

            X_sample_r = np.delete(Xj, r)

            return(X_sample_r)

        
        replicas_estimador = []

        for r in range(0, len(X)):

            replicas_estimador.append( np.std( Jacknife_sample(X , r) ) )

    
        n = len(X)

        sesgo = (n-1) * ( np.mean( replicas_estimador ) - np.std(X) )

        estimacion_sesgo_corregido = np.std(X) - sesgo

        standard_error = np.sqrt( ((n-1)/n) * sum( (replicas_estimador - np.mean( replicas_estimador ))**2 ) )


#######################################################

    if estimador == 'quantile':

        def Jacknife_sample(X , r):

            X_sample_r = np.delete(X, r)

            return(X_sample_r)

        
        replicas_estimador = []

        for r in range(0, len(X)):

            replicas_estimador.append( np.quantile( Jacknife_sample(X , r) , q) )

    
        n = len(X)

        sesgo = (n-1) * ( np.mean( replicas_estimador ) - np.quantile(X , q) )

        estimacion_sesgo_corregido = np.quantile(X , q) - sesgo

        standard_error = np.sqrt( ((n-1)/n) * sum( (replicas_estimador - np.mean( replicas_estimador ))**2 ) )

#######################################################

    if estimador == 'kurtosis':

        import scipy

        from scipy.stats import  kurtosis

        def Jacknife_sample(X , r):

            X_sample_r = np.delete(X, r)

            return(X_sample_r)

        
        replicas_estimador = []

        for r in range(0, len(X)):

            replicas_estimador.append( kurtosis(X) )

    
        n = len(X)

        sesgo = (n-1) * ( np.mean( replicas_estimador ) - kurtosis(X) )

        estimacion_sesgo_corregido = kurtosis(X) - sesgo

        standard_error = np.sqrt( ((n-1)/n) * sum( (replicas_estimador - np.mean( replicas_estimador ))**2 ) )


#######################################################
     
#######################################################

    if estimador == 'skewness':

        import scipy

        from scipy.stats import  skew

        def Jacknife_sample(X , r):

            X_sample_r = np.delete(X, r)

            return(X_sample_r)

        
        replicas_estimador = []

        for r in range(0, len(X)):

            replicas_estimador.append( skew(X) )

    
        n = len(X)

        sesgo = (n-1) * ( np.mean( replicas_estimador ) - skew(X) )

        estimacion_sesgo_corregido = skew(X) - sesgo

        standard_error = np.sqrt( ((n-1)/n) * sum( (replicas_estimador - np.mean( replicas_estimador ))**2 ) )


#######################################################
    
    return(sesgo, estimacion_sesgo_corregido, standard_error)

In [3]:
def Jacknife(Variable , estimator_function, q=0.75):

#######################################################

    def Jacknife_sample(X , r):

            X_sample_r = np.delete(X, r)

            return(X_sample_r)


    if estimator_function == np.quantile :  estimation = estimator_function(Variable , q=q)

    else :  estimation = estimator_function(Variable)

        
    replicas_estimador = []

    for r in range(0, len(Variable)):

        if estimator_function == np.quantile :  Jack_estimation = estimator_function( Jacknife_sample(Variable, r) , q=q )

        else :  Jack_estimation = estimator_function( Jacknife_sample(Variable, r) )

        replicas_estimador.append( Jack_estimation )

    
    n = len(Variable)

    sesgo = (n-1) * ( np.mean( replicas_estimador ) - estimation )

    estimacion_sesgo_corregido = estimation - sesgo

    standard_error = np.sqrt( ((n-1)/n) * sum( (replicas_estimador - np.mean( replicas_estimador ))**2 ) )


#######################################################
    
    return(sesgo, estimacion_sesgo_corregido, standard_error)

In [4]:
np.random.seed(123)

X = np.random.normal(loc=10, scale=15, size=50)

In [5]:
np.median(X)

8.23916733155832

In [6]:
sesgo, estimacion_sesgo_corregido, standard_error = Jacknife(X , estimator_function=np.median , q=0.75)

In [7]:
sesgo

8.704148513061227e-14

In [8]:
estimacion_sesgo_corregido

8.239167331558233

In [9]:
standard_error

2.381386940718188

In [10]:
np.mean(X)

10.199071616256777

In [11]:
sesgo, estimacion_sesgo_corregido, standard_error = Jacknife(X , estimator_function=np.mean)

In [12]:
sesgo

-8.704148513061227e-14

In [13]:
estimacion_sesgo_corregido

10.199071616256864

In [14]:
standard_error

2.5491917443460235

In [15]:
np.std(X)

17.844342210422187

In [16]:
sesgo, estimacion_sesgo_corregido, standard_error = Jacknife(X , estimator_function=np.std)

In [17]:
sesgo

-0.26116994703598806

In [18]:
estimacion_sesgo_corregido

18.105512157458175

In [19]:
standard_error

1.6795955569730596

In [20]:
np.quantile(X , q=0.75)

23.835596841535178

In [21]:
sesgo, estimacion_sesgo_corregido, standard_error = Jacknife(X , np.quantile, q=0.75)

In [22]:
sesgo

-0.14962568429344003

In [23]:
estimacion_sesgo_corregido

23.98522252582862

In [24]:
standard_error

0.9375849844484524

In [25]:
import scipy

from scipy.stats import  kurtosis

In [26]:
kurtosis(X)

-0.37420768292897266

In [27]:
sesgo, estimacion_sesgo_corregido, standard_error = Jacknife(X , estimator_function=kurtosis)

In [28]:
sesgo

-0.03624894442748883

In [29]:
estimacion_sesgo_corregido

-0.33795873850148384

In [30]:
standard_error

0.3609496287814513

In [31]:
import scipy

from scipy.stats import  skew

In [32]:
skew(X)


0.025587358812510053

In [33]:
sesgo, estimacion_sesgo_corregido, standard_error = Jacknife(X , estimator_function=skew)

In [34]:
sesgo

0.021610720628010085

In [35]:
estimacion_sesgo_corregido

0.0039766381844999685

In [36]:
standard_error

0.27115376266443825

# Bootstrap

In [37]:
def Bootstrap(Variable , B, estimator_function, q=0.75, random_seed=123 ):

#######################################################

    np.random.seed(random_seed)

#######################################################


    def Bootstrap_sample(X):

            from sklearn.utils import resample

            sample = resample( X, n_samples=len(X))

            return sample


    if estimator_function == np.quantile : estimation = estimator_function(Variable , q)
            
    else:  estimation = estimator_function(Variable)


    replicas_estimador = []

    for b in range(0, B):

        if estimator_function == np.quantile : estimation_Boot = estimator_function(Bootstrap_sample(Variable) , q=q)
            
        else:  estimation_Boot = estimator_function(Bootstrap_sample(Variable))

        replicas_estimador.append( estimation_Boot )

    
    sesgo = np.mean( replicas_estimador ) - estimation

    estimacion_sesgo_corregido = estimation - sesgo

    standard_error = np.std( replicas_estimador ) 


#######################################################

    return sesgo , estimacion_sesgo_corregido , standard_error, replicas_estimador

In [38]:
np.median(X)

8.23916733155832

In [39]:
sesgo , estimacion_sesgo_corregido , standard_error, replicas_estimador = Bootstrap(X , B=20000, estimator_function=np.median)

In [40]:
sesgo

0.4499295142913571

In [41]:
estimacion_sesgo_corregido

7.789237817266963

In [42]:
standard_error

3.7580290713770914

In [43]:
sesgo , estimacion_sesgo_corregido , standard_error, replicas_estimador = Bootstrap(X , B=20000, estimator_function=np.mean)

In [44]:
np.mean(X)

10.199071616256777

In [45]:
sesgo

0.02044617187141995

In [46]:
estimacion_sesgo_corregido

10.178625444385357

In [47]:
standard_error

2.5121071650105433

In [48]:
sesgo , estimacion_sesgo_corregido , standard_error, replicas_estimador = Bootstrap(X , B=20000, estimator_function=np.std)

In [49]:
np.std(X)

17.844342210422187

In [50]:
replicas_estimador[0:10]

[17.290069166352062,
 17.734700491052042,
 18.091905056508867,
 19.296946848609828,
 16.371718686759674,
 17.889036386608268,
 16.84496188781967,
 18.60024334635148,
 20.188328416078278,
 19.38523325913237]

In [51]:
sesgo

-0.24900263565758252

In [52]:
estimacion_sesgo_corregido

18.09334484607977

In [53]:
standard_error

1.6154115156505968

In [54]:
sesgo , estimacion_sesgo_corregido , standard_error, replicas_estimador = Bootstrap(X , B=20000, estimator_function=np.quantile, q=0.75)

In [55]:
np.quantile(X, q=0.75)

23.835596841535178

In [56]:
replicas_estimador[0:5]

[19.843817079990643,
 23.835596841535178,
 16.093460800379468,
 25.060808468183154,
 26.898336776063715]

In [57]:
sesgo

-1.006032620023607

In [58]:
estimacion_sesgo_corregido

24.841629461558785

In [59]:
standard_error

3.4665427821179793

In [144]:
skew(X)

0.025587358812510053

In [133]:
sesgo , estimacion_sesgo_corregido , standard_error, replicas_estimador = Bootstrap(X , B=20000, estimator_function=skew)

In [134]:
sesgo

0.019212954042191917

In [135]:
estimacion_sesgo_corregido

0.006374404770318136

In [136]:
standard_error

0.253427740069774

In [145]:
kurtosis(X)

-0.37420768292897266

In [140]:
sesgo , estimacion_sesgo_corregido , standard_error, replicas_estimador = Bootstrap(X , B=20000, estimator_function=kurtosis)

In [141]:
sesgo

-0.022470288214714473

In [142]:
estimacion_sesgo_corregido

-0.3517373947142582

In [143]:
standard_error

0.35103721258189574

## Intervalos de confianza cuantil-bootstrap

### Intervalo para un parametro (media, varianza, mediana, cuantiles, asimetria, curtosis)

In [60]:
def cuantil_boot_interval(Variable1, Variable2, alpha, estimator , B,  q=0.75, random_seed=123):

    from itertools import chain 

######################################################################

    np.random.seed(random_seed)

######################################################################

    def Bootstrap_sample(Variable):

        from sklearn.utils import resample

        sample = resample( Variable, n_samples=len(Variable))

        return sample

######################################################################

    replicas_estimador = []


    if estimator == 'mean':

        for b in range(0, B):

            replicas_estimador.append( np.mean( Bootstrap_sample(Variable1) ) )

        estimation = np.mean(Variable1)

######################################################################
#     
    if estimator == 'median':

        for b in range(0, B):

            replicas_estimador.append( np.median( Bootstrap_sample(Variable1) ) )

        estimation = np.median(Variable1)

######################################################################

    if estimator == 'std':

        for b in range(0, B):

            replicas_estimador.append( np.std( Bootstrap_sample(Variable1) ) )

        estimation = np.std(Variable1)

######################################################################

    if estimator == 'skewness':

        from scipy.stats import skew

        for b in range(0, B):

            replicas_estimador.append( skew( Bootstrap_sample(Variable1) ) )

        estimation = skew(Variable1)

######################################################################

    if estimator == 'kurtosis':

        from scipy.stats import kurtosis

        for b in range(0, B):

            replicas_estimador.append( kurtosis( Bootstrap_sample(Variable1) ) )

        estimation = kurtosis(Variable1)

######################################################################

    if estimator == 'quantile':

        for b in range(0, B):

            replicas_estimador.append( np.quantile( Bootstrap_sample(Variable1) , q=q ) )

        estimation = np.quantile(Variable1 , q=q)

######################################################################

    if estimator == 'proportion': # Variable1 debe ser una variable categorica **binaria**.

        for b in range(0, B):

            replicas_estimador.append( np.mean( Bootstrap_sample(Variable1) ) )

        estimation = np.mean(Variable1)



######################################################################

######################################################################

    replicas_estimador_1 , replicas_estimador_2 = [] , []

    
    if estimator == 'mean_diff':

        for b in range(0, B):

            replicas_estimador_1.append( np.mean( Bootstrap_sample(Variable1) ) )

            replicas_estimador_2.append( np.mean( Bootstrap_sample(Variable2) ) )
        
        replicas_estimador_diff = np.array(replicas_estimador_1) - np.array(replicas_estimador_2)

        estimation = np.mean(Variable1) - np.mean(Variable2)      

######################################################################

    if estimator == 'median_diff':

        for b in range(0, B):

            replicas_estimador_1.append( np.median( Bootstrap_sample(Variable1) ) )

            replicas_estimador_2.append( np.median( Bootstrap_sample(Variable2) ) )
        
        replicas_estimador_diff = np.array(replicas_estimador_1) - np.array(replicas_estimador_2)

        estimation = np.median(Variable1) - np.median(Variable2)      

######################################################################

    if estimator == 'std_diff':

        for b in range(0, B):

            replicas_estimador_1.append( np.std( Bootstrap_sample(Variable1) ) )

            replicas_estimador_2.append( np.std( Bootstrap_sample(Variable2) ) )
        
        replicas_estimador_diff = np.array(replicas_estimador_1) - np.array(replicas_estimador_2)

        estimation = np.std(Variable1) - np.std(Variable2)      

######################################################################

    if estimator == 'quantile_diff':

        for b in range(0, B):

            replicas_estimador_1.append( np.quantile( Bootstrap_sample(Variable1), q=q) )

            replicas_estimador_2.append( np.quantile( Bootstrap_sample(Variable2), q=q ) )
        
        replicas_estimador_diff = np.array(replicas_estimador_1) - np.array(replicas_estimador_2)

        estimation = np.quantile(Variable1, q=q) - np.quantile(Variable2, q=q)      

######################################################################

    if estimator == 'skewness_diff':

        from scipy.stats import skew

        for b in range(0, B):

            replicas_estimador_1.append( skew( Bootstrap_sample(Variable1) ) )

            replicas_estimador_2.append( skew( Bootstrap_sample(Variable2) ) )
        
        replicas_estimador_diff = np.array(replicas_estimador_1) - np.array(replicas_estimador_2)

        estimation = skew(Variable1) - skew(Variable2)      

######################################################################

    if estimator == 'kurtosis_diff':

        from scipy.stats import kurtosis

        for b in range(0, B):

            replicas_estimador_1.append( kurtosis( Bootstrap_sample(Variable1) ) )

            replicas_estimador_2.append( kurtosis( Bootstrap_sample(Variable2) ) )
        
        replicas_estimador_diff = np.array(replicas_estimador_1) - np.array(replicas_estimador_2)

        estimation = kurtosis(Variable1) - kurtosis(Variable2)   

######################################################################

    if estimator == 'proportion_diff': # Variable1 y Variable2 deben ser variables categoricas **binarias**.

        for b in range(0, B): 

            replicas_estimador_1.append( np.mean( Bootstrap_sample(Variable1) ) )

            replicas_estimador_2.append( np.mean( Bootstrap_sample(Variable2) ) )
        
        replicas_estimador_diff = np.array(replicas_estimador_1) - np.array(replicas_estimador_2)

        estimation = np.mean(Variable1) - np.mean(Variable2)       

######################################################################
######################################################################

    if estimator in ['mean','median','std','quantile','kurtosis','skewness','proportion']:

        L1_1 = np.quantile( replicas_estimador  , q=alpha/2)

        L2_1 = np.quantile( replicas_estimador  , q=1-alpha/2)

        interval = [L1_1 , L2_1]

######################################################################

    if estimator in ['mean_diff','median_diff','std_diff','quantile_diff','kurtosis_diff','skewness_diff', 'proportion_diff']:
   
        L1_2 = np.quantile( replicas_estimador_diff  , q=alpha/2)

        L2_2 = np.quantile( replicas_estimador_diff  , q=1-alpha/2)

        interval = [L1_2 , L2_2]

######################################################################
######################################################################

    return interval , estimation

### Para la media

In [61]:
interval , estimation = cuantil_boot_interval(Variable1=X, Variable2='no', alpha=0.05, estimator='mean' , B=20000, random_seed=123)

In [62]:
interval

[5.274553626813773, 15.146375204290525]

In [63]:
estimation

10.199071616256777

In [64]:
def CI_Mean(Variable , alpha=0.05):

    n = len(Variable)

    t_alpha_medios = scipy.stats.t.ppf( 1 - alpha/2 , df=n-1)

    X_mean = Variable.mean()

    X_cuasi_var = Variable.std()**2 
    
    # std() esta definida por defecto como la cuasi-desviacion-tipica

    L1 = X_mean - t_alpha_medios * np.sqrt(X_cuasi_var/n)

    L2 = X_mean + t_alpha_medios * np.sqrt(X_cuasi_var/n)

    interval = [L1 , L2]

    return interval , X_mean

In [65]:
interval , X_mean = CI_Mean(X , alpha=0.05)

In [66]:
interval

[5.127765678327374, 15.27037755418618]

### Para la desviacion tipica

In [67]:
interval , estimation = cuantil_boot_interval(Variable1=X, Variable2='no', alpha=0.05, estimator='std' , B=20000, random_seed=123)

In [68]:
interval

[14.427214668666169, 20.756590926218593]

In [69]:
estimation

17.844342210422187

In [70]:
def CI_Variance(Variable , alpha=0.05):

    n = len(Variable)

    chi_alpha_medios = scipy.stats.chi2.ppf( 1 - alpha/2 , df=n-1)

    chi_1_alpha_medios = scipy.stats.chi2.ppf(alpha/2 , df=n-1)

    X_cuasi_var = Variable.std()**2 

    X_var = ( (n-1)/n )*X_cuasi_var
    
    # std() esta definida por defecto como la cuasi-desviacion-tipica

    L1 =  (n*X_var) / chi_alpha_medios

    L2 = (n*X_var) / chi_1_alpha_medios

    interval = [L1 , L2]

    return interval , X_var 

In [71]:
interval , X_var = CI_Variance(Variable=X , alpha=0.05)

In [72]:
interval

[222.18841684285198, 494.45882436956106]

In [73]:
X_var

312.0521379442019

In [74]:
np.sqrt(interval)

array([14.90598594, 22.23643012])

In [75]:
X_var

312.0521379442019

### Para la mediana

In [76]:
interval , estimation = cuantil_boot_interval(Variable1=X, Variable2='no', alpha=0.05, estimator='median' , B=20000, random_seed=123)

In [77]:
interval

[2.4028635430970975, 15.99323919612726]

In [78]:
estimation

8.23916733155832

### Para la asimetria

In [79]:
interval , estimation = cuantil_boot_interval(Variable1=X, Variable2='no', alpha=0.05, estimator='skewness' , B=20000, random_seed=123)

In [80]:
interval

[-0.43998686459545305, 0.5624134382838676]

In [81]:
estimation

0.025587358812510053

### Para la curtosis

In [82]:
interval , estimation = cuantil_boot_interval(Variable1=X, Variable2='no', alpha=0.05, estimator='kurtosis' , B=20000, random_seed=123)

In [83]:
interval 

[-0.9952663071794702, 0.38601833503542765]

In [84]:
estimation

-0.37420768292897266

### Para los cuantiles

In [85]:
interval , estimation = cuantil_boot_interval(Variable1=X, Variable2='no', alpha=0.05, estimator='quantile' , B=20000, random_seed=123, q=0.75)

In [86]:
interval

[15.078835764997024, 28.98904388058301]

In [87]:
estimation

23.835596841535178

### Para la proporción

In [148]:
X_dummy = np.random.uniform(low=0 , high=1, size=50).round()

In [149]:
interval , estimation = cuantil_boot_interval(Variable1=X_dummy, Variable2='no', alpha=0.05, estimator='proportion' , B=20000, random_seed=123)

In [150]:
interval

[0.46, 0.74]

In [151]:
estimation

0.6

## Intervalo para la diferencia de parametros (medias, medianas, varianzas, cuantiles, curtosos, asimetrias)

In [88]:
np.random.seed(123)

X_1 = np.random.normal(loc=10, scale=15, size=50)

X_2 = np.random.normal(loc=13, scale=15, size=100)

### Para la diferencia de medias

In [89]:
interval , estimation = cuantil_boot_interval(Variable1=X_1, Variable2=X_2, alpha=0.05, estimator='mean_diff' , B=20000,  q=0.75, random_seed=123)

In [90]:
interval

[-9.788335434742939, 1.810648408844556]

In [91]:
estimation

-4.001757444976697

In [92]:
def CI_Mean_Diference(Variable1 , Variable2 , alpha=0.05):

    X1 = Variable1
    X2 = Variable2

    n1 = len(X1) 
    n2 = len(X2)   

    X1_mean = X1.mean()
    X2_mean = X2.mean()

    X1_cuasi_var = X1.std()**2 
    X2_cuasi_var = X2.std()**2 

    X1_var = ( (n1-1)/n1 )*X1_cuasi_var
    X2_var = ( (n2-1)/n2 )*X2_cuasi_var

    v = ( X1_var/n1 + X2_var/n2 )**2 / ( (X1_var/n1)**2 / (n1-1)  + (X2_var/n2)**2 / (n2-1)  )

    t_alpha_medios = scipy.stats.chi.ppf( 1 - alpha/2 , df=v)
  
  
    L1 =  (X1_mean - X2_mean) - t_alpha_medios * np.sqrt(X1_var/n1 + X2_var/n2)

    L2 =  (X1_mean - X2_mean) + t_alpha_medios * np.sqrt(X1_var/n1 + X2_var/n2)

    interval = [L1 , L2]

    return interval , (X1_mean - X2_mean) 

In [93]:
interval , estimation = CI_Mean_Diference(Variable1=X_1 , Variable2=X_2 , alpha=0.05)

In [94]:
interval

[-35.535058586780195, 27.531543696826805]

In [95]:
estimation

-4.001757444976697

### Para la diferencia de medianas

In [96]:
interval , estimation = cuantil_boot_interval(Variable1=X_1, Variable2=X_2, alpha=0.05, estimator='median_diff' , B=20000,  q=0.75, random_seed=123)

In [97]:
interval

[-13.881285812364158, 3.1918214145850397]

In [98]:
estimation

-6.812252818435496

### Para la diferencia de desviaciones tipicas

In [99]:
interval , estimation = cuantil_boot_interval(Variable1=X_1, Variable2=X_2, alpha=0.05, estimator='std_diff' , B=20000,  q=0.75, random_seed=123)

In [100]:
interval

[-1.440066276291272, 5.750492255040537]

In [101]:
estimation

2.297921770341434

### Para la diferencia de cuantiles

In [102]:
interval , estimation = cuantil_boot_interval(Variable1=X_1, Variable2=X_2, alpha=0.05, estimator='quantile_diff' , B=20000,  q=0.75, random_seed=123)

In [103]:
interval

[-12.249938782339909, 4.819498365231816]

In [104]:
estimation

-1.2929002407485939

### Para la diferencia de kurtosis

In [105]:
interval , estimation = cuantil_boot_interval(Variable1=X_1, Variable2=X_2, alpha=0.05, estimator='kurtosis_diff' , B=20000,  q=0.75, random_seed=123)

In [106]:
interval

[-0.504742036075702, 1.1320430654741647]

In [107]:
estimation

0.2817371126835462

### Para la diferencia de skewness

In [108]:
interval , estimation = cuantil_boot_interval(Variable1=X_1, Variable2=X_2, alpha=0.05, estimator='skewness_diff' , B=20000,  q=0.75, random_seed=123)

In [109]:
interval

[-0.5445430937619894, 0.6394836839729549]

In [110]:
estimation

0.019079758603481104

### Para la diferencia de proporciones

In [181]:
np.random.seed(123)

X_dummy_1 = np.random.uniform(low=0 , high=1, size=40).round()

X_dummy_2 = np.random.uniform(low=0 , high=1, size=300).round()

In [182]:
interval , estimation = cuantil_boot_interval(Variable1=X_dummy_1, Variable2=X_dummy_2, alpha=0.05, estimator='proportion_diff' , B=20000,  q=0.75, random_seed=123)

In [183]:
interval

[-0.22499999999999998, 0.10666666666666663]

In [184]:
estimation

-0.06

## Intervalos BCa

In [111]:
def BCa_boot_interval(Variable, estimator_function, B, alpha, random_seed, q=0.75):

    np.random.seed(random_seed)


    def Bootstrap_sample(Variable):

        from sklearn.utils import resample

        sample = resample( Variable, n_samples=len(Variable))

        return sample


    def Jacknife_sample(X , r):

            X_sample_r = np.delete(X, r)

            return(X_sample_r)



    z_1_alpha_medios = scipy.stats.norm.ppf(q=alpha/2, loc=0, scale=1)

    z_alpha_medios = scipy.stats.norm.ppf(q=1-alpha/2, loc=0, scale=1)

##################################################################################

    if estimator_function == np.quantile:  estimation = estimator_function(Variable, q=q)

    else :  estimation = estimator_function(Variable)

    replicas_boot_estimador = []

    for b in range(0, B):

        replicas_boot_estimador.append( np.mean( Bootstrap_sample(Variable) ) )


    replicas_boot_estimador = np.array(replicas_boot_estimador)


    rho = sum( replicas_boot_estimador <= estimation ) / B

    z_0 = scipy.stats.norm.ppf(q=rho, loc=0, scale=1)

    ###################################

    replicas_jack_estimador = []

    for r in range(0, len(Variable)):

        if estimator_function == np.quantile :  Jack_estimation = estimator_function( Jacknife_sample(Variable, r) , q=q )

        else :  Jack_estimation = estimator_function( Jacknife_sample(Variable, r) )

        replicas_jack_estimador.append( Jack_estimation )

    replicas_jack_estimador = np.array(replicas_jack_estimador)


    a_numerator = sum( (np.mean(replicas_jack_estimador) - replicas_jack_estimador )**3 )

    a_denominator = sum( (np.mean(replicas_jack_estimador) - replicas_jack_estimador )**2 )

    a_denominator = 6*a_denominator**(3/2)

    a = a_numerator / a_denominator


    x_1 = z_0 + (z_0 + z_1_alpha_medios)/(1-a*(z_0 + z_1_alpha_medios))

    x_2 = z_0 + (z_0 + z_alpha_medios)/(1-a*(z_0 + z_alpha_medios))


    alpha_1 = scipy.stats.norm.cdf(x=x_1, loc=0, scale=1)

    alpha_2 = scipy.stats.norm.cdf(x=x_2, loc=0, scale=1)


#########################################################################################################

    L1 = np.quantile( replicas_boot_estimador  , q=alpha_1)

    L2 = np.quantile( replicas_boot_estimador  , q=alpha_2)

    interval = [L1,L2]

    return interval , estimation

In [112]:
interval , estimation = BCa_boot_interval(Variable=X, estimator_function=np.mean, B=200000, alpha=0.05, random_seed=123)

In [113]:
interval

[5.262399300501412, 15.148561429340479]

In [114]:
estimation

10.199071616256777

In [115]:
interval , estimation = BCa_boot_interval(Variable=X, estimator_function=np.median, B=200000, alpha=0.05, random_seed=123)

In [116]:
interval

[1.1580146455412295, 11.197676226492618]

In [117]:
estimation

8.23916733155832

In [118]:
interval , estimation = BCa_boot_interval(Variable=X, estimator_function=np.std, B=200000, alpha=0.05, random_seed=123)

In [119]:
interval

[20.774957571314236, 21.430372642336394]

In [120]:
estimation

17.844342210422187

In [121]:
from scipy.stats import kurtosis

In [122]:
interval , estimation = BCa_boot_interval(Variable=X, estimator_function=kurtosis , B=200000, alpha=0.05, random_seed=123, q=0.5)

In [123]:
interval

[-0.6003612573539066, -0.6003597491211978]

In [124]:
estimation

-0.37420768292897266

In [125]:
from scipy.stats import skew

In [126]:
interval , estimation = BCa_boot_interval(Variable=X, estimator_function=skew , B=200000, alpha=0.05, random_seed=123, q=0.5)

In [127]:
interval

[-0.6003612573539066, -0.6003362491509903]

In [128]:
estimation

0.025587358812510053

In [129]:
interval , estimation = BCa_boot_interval(Variable=X, estimator_function=np.quantile , B=200000, alpha=0.05, random_seed=123, q=0.6)

In [130]:
interval

[14.014587566665833, 21.42972709983565]

In [131]:
estimation

14.584180220264427

In [132]:
interval , estimation = BCa_boot_interval(Variable=X, estimator_function=np.quantile , B=200000, alpha=0.05, random_seed=123, q=0.75)

C:\Users\Usuario\AppData\Local\Temp\ipykernel_8672\801661732.py:71: RuntimeWarning: invalid value encountered in double_scalars
  x_1 = z_0 + (z_0 + z_1_alpha_medios)/(1-a*(z_0 + z_1_alpha_medios))
C:\Users\Usuario\AppData\Local\Temp\ipykernel_8672\801661732.py:73: RuntimeWarning: invalid value encountered in double_scalars
  x_2 = z_0 + (z_0 + z_alpha_medios)/(1-a*(z_0 + z_alpha_medios))


ValueError: Quantiles must be in the range [0, 1]

## Contrastes Bootstrap

In [ ]:
def boot_test_mean(Variable, mean_0, H1_type, B, alpha):

    n = len(Variable)

    from sklearn.utils import resample

   
    def Bootstrap_sample(X):

        sample = resample( X, n_samples=len(X))

        return sample

    def statistic_test(X):

        statistic = (X.mean() - mean_0 )/np.sqrt(np.std(X)**2 / n)

        return statistic



    replicas_statistic_test = []


    for b in range(0, B):

        replicas_statistic_test.append( statistic_test( Bootstrap_sample(Variable) ) )



    #############################################################

    if H1_type == 'greater':  
        
        ASL = sum(replicas_statistic_test >= statistic_test(Variable)) / B

        if ASL <= alpha : 
            
            result = 'Reject H0: mean = mean_0  in favor of  H1: mean > mean_0  -->  Accept H1: mean > mean_0'

        else : 
            
            result = 'Not Reject H0: mean = mean_0  in favor of  H1: mean > mean_0  -->  Not Accept H1: mean > mean_0'


    if H1_type == 'less':
        
        ASL = sum(replicas_statistic_test <= statistic_test(Variable)) / B

        if ASL <= alpha : 
            
            result = 'Reject H0: mean = mean_0  in favor of  H1: mean < mean_0  -->  Accept H1: mean < mean_0'

        else : 
            
            result = 'Not Reject H0: mean = mean_0  in favor of  H1: mean < mean_0  -->  Not Accept H1: mean < mean_0'


    if H1_type == 'two.sided':
        
        ASL = sum(replicas_statistic_test >= abs(statistic_test(Variable))) / B   +   sum(replicas_statistic_test <= -abs(statistic_test(Variable))) / B 

        if ASL <= alpha : 
            
            result = 'Reject H0: mean = mean_0  in favor of  H1: mean =/= mean_0  -->  Accept H1: mean =/= mean_0'

        else : 
            
            result = 'Not Reject H0: mean = mean_0  in favor of  H1: mean =/= mean_0  -->  Not Accept H1: mean =/= mean_0'

    #############################################################

    return ASL , result

In [ ]:
import pandas as pd

In [ ]:
Data = pd.read_csv('House_Price_Regression.csv')

In [ ]:
Price = Data.price

In [ ]:
Price.mean()

2085829.8734908137

In [ ]:
boot_test_mean(Variable=Price, mean_0=1000, H1_type='greater', B=20000 , alpha=0.05)

(0.52615,
 'Not Reject H0: mean = mean_0  in favor of  H1: mean > mean_0  -->  Not Accept H1: mean > mean_0')

In [ ]:
boot_test_mean(Variable=Price, mean_0=1000, H1_type='less', B=20000 , alpha=0.05)

(0.47545,
 'Not Reject H0: mean = mean_0  in favor of  H1: mean < mean_0  -->  Not Accept H1: mean < mean_0')

In [ ]:
boot_test_mean(Variable=Price, mean_0=1000, H1_type='two.sided', B=20000 , alpha=0.05)

(0.5215,
 'Not Reject H0: mean = mean_0  in favor of  H1: mean =/= mean_0  -->  Not Accept H1: mean =/= mean_0')

## Contrastes de hipotesis por permutaciones

El test de permutaciones es un test de significancia estadística para el estudio de diferencias entre grupos. Fue desarrollado por Ronald Fisher y E.J.G. Pitman en 1930. La distribución del estadístico estudiado (media, mediana...) se obtiene calculando el valor de dicho estadístico para todas las posibles reorganizaciones de las observaciones en los distintos grupos. Dado que implica calcular todas las posibles situaciones, se trata de un test exacto.



La filosofia de los test de permutaciones es la siguiente:



Se tiene un conjunto de individuos, los cuales se distribuyen aleatoriamente en dos grupos, $T$ y $C$, de tamaños  $n_t$  y  $n_c$



Se aplica un experimento, que generalmente consiste en aplicar un tratamiento sobre un grupo (grupo tratamiento $T$) y no aplicarlo sobre el otro grupo (grupo control $C$).



Se mide una variable de interés $X$ en cada uno de los grupos:

$$ X^t = (x^t_{1},...,x^t_{n_t }) $$
$$ X^c = (x^c_{1},...,x^c_{n_c }) $$

Donde: 

$x^t_{i}$ es el valor de $X$ para el $i$-esimo individuo del grupo $T$

$x^c_{i}$ es el valor de $X$ para el $i$-esimo individuo del grupo $C$




Se calcula un estadistico $\widehat{\theta}$ de $X$ en los grupos $\Rightarrow \widehat{\theta}_t = f(X^t), \widehat{\theta}_c = f(X^c)$

Estadisticos habituales son la media y la mediana.



Se desea  determinar si existen diferencias significativas entre el valor del estadistico para cada grupo. 

Para ello se considera que si la diferencia observada en valor absoluto ($\mid D_{obs} \mid = \mid \widehat{\theta}_t - \widehat{\theta}_c \mid$ ) es 
mayor que la diferencia esperada por la asignación aleatoria de los grupos, en valor absoluto.
 

En términos estadísticos, la hipótesis nula ( $H_0$ ) es que ambas muestras pertenecen a la misma distribución y que, las diferencias observadas, son debidas únicamente a variaciones causadas por el reparto aleatorio en grupos. 



El test de permutación permite identificar si hay evidencias en contra de esta hipótesis. Para ello los pasos a seguir son:

1) Calcular la diferencia observada del estadistico para los dos grupos  

 $$\mid D_{obs} \mid = \mid \widehat{\theta}_t - \widehat{\theta}_c \mid$$



2) Todas las observaciones se unen sin tener en cuenta el grupo al que pertenecen. $(X^t , X^c)$



3) Se generan $B$ pares de muestras  sin reemplazamiento  de $(X^t , X^c)$ , de tamaño $n_t$ y $n_c$ respectivamente :

$$(\mathcal{X}^t_1 , \mathcal{X}^c_1), (\mathcal{X}^t_2 , \mathcal{X}^c_2),\dots, (\mathcal{X}^t_B , \mathcal{X}^c_B)$$


   $\mathcal{X}^g_i$ es la muestra aleatoria $i$-esima de tamaño n_g extraida de  $(X^t , X^c)$ , para $g=t,c$ , $i=1,...,B$
   
  



4) Para cada combinación, calcular la diferencia entre medias  (difcalculada) . El conjunto de valores obtenidos forman la distribución exacta de las posibles diferencias siendo cierta la hipótesis nula de que los dos grupos proceden de la misma población. A esta distribución, se le conoce como permutation distribution of the mean difference.

5) Calcular el p-value de dos colas como la proporción de combinaciones en las que, el valor absoluto de la diferencia calculada, es mayor o igual al valor absoluto de la diferencia observada. La idea es que si esta proporcion no es suficientemente pequeña, habrá evidencias de que las diferencias observadas entre los grupos son fruto del azar involucrado al generar los grupos y no del tratamiento.  Si en cambio esta proporcion es suficientemente pequeña, habra evidencias de que las doferencias observadas son causadas por el tratamiento, y no por la configuracion aleatoria de los grupos. Pero hay que delimitar un punto a partir del cual considerar esta proporcion como suficientmiente pequeña o no.

6) Si el p-value calculado en 5) es mayor que cierto limite $\alpha$ , no se podra rechazar H_0 en favor de H_1 puesto que se entiende que no hay suficiente evidencia. En caso de p-value menor que \alpha se acepta H_1 dada la evidencia dispobible. Normalmente se considera \alpha = 0.05 , 0.1 0 o 0.01 , en funcion del nivel de evidencia exigido. A mayor alpha, menos evidencia se exige para aceptar H_1 , cuando menor alpha mayor evidencia exigida para aceptar H_1

En este ejemplo se emplea como estadístico la media pero podría ser cualquier otro (mediana, varianza...)

Para que el test sea exacto deberia tomarse $B= \dfrac{(n_t+n_c)!}{n_t! \hspace{0.1cm} n_c!}$ , que es el numero de diferentes combinaciones  para distribuir $n_t + n_c$  observaciones en dos grupos de tamaños $n_t$ y $n_c$ 

Y ademas las B pares de muestras extraidas deberian de ser distintos (sin que importe el orden de los elementos, es decir dos muestras con los mismos elementos pero en diferente orden son consideradas iguales).

El problema de este enfoque exacto es que para tamaños moderados de $n_t + n_c$ se llegan a valores inasumibles de $B= \dfrac{(n_t+n_c)!}{n_t! \hspace{0.1cm} n_c!}$

Por ejemplo:

In [ ]:
from math import factorial

In [ ]:
factorial(30) / (factorial(20)*factorial(10))

30045015.0

In [ ]:
factorial(30) / (factorial(15)*factorial(15))

155117520.0

La única condición necesaria para un test de permutaciones se conoce como exchangeability, según la cual, todas las posibles permutaciones tienen la misma probabilidad de ocurrir siendo cierta la hipótesis nula.

Los test de permutación son test de significancia y, por lo tanto, se emplean para calcular p-values. Es importante tener en cuenta que, las conclusiones extraídas, solo son aplicables a diseños de tipo experimental. Estudios en los que, tras haber elegido los sujetos, se realiza una asignación aleatoria a los diferentes grupos.

Nota: si bien se utiliza el término permutación, se trata realmente de combinaciones, ya que el orden de las observaciones dentro de cada grupo no importa.

Simulación de Monte Carlo

En los test de permutación, el p-value obtenido es exacto ya que se calculan todas las posibles combinaciones de las observaciones. Esto resulta complicado o imposible cuando el tamaño muestral es mediano o grande. Por ejemplo, existen más de 155 millones combinaciones distintas para distribuir 30 observaciones en dos grupos de 15 cada uno.



(m+n)!m!n!=(15+15)!15!15!=155117520
 


In [ ]:
# Número total de combinaciones
# ==============================================================================
from math import factorial
print(f"Total combinaciones = {factorial(15 + 15) / (factorial(15)*factorial(15))}")



Total combinaciones = 155117520.0
La simulación de Monte Carlo o randomization test consiste en emplear únicamente una muestra aleatoria de todas las posibles combinaciones, evitando así tener que calcularlas todas. Al no calcularse todas las posibles combinaciones, no es un test exacto sino aproximado.

El método de Monte Carlo no es insesgado. Una de las correcciones más empleadas es la propuesta por Davison and Hinkley):

pvalue=r+1k+1
 
siendo r el número de combinaciones igual o más extremas que el estadístico observado y k el número de combinaciones empleadas en la simulación.

A pesar de que la corrección es mínima cuando el número de simulaciones es alto, presenta la ventaja de que, si el valor observado es mayor que cualquiera de los calculados, el p-value es muy bajo pero no cero. Por facilidades de cálculo, se suele emplear un número de permutaciones terminado en 9 (4999, 9999).

¿Cuántas permutaciones (combinaciones) son necesarias?

No hay un número predefinido de combinaciones que asegure buenos resultados. Una forma de guiar la elección es tener en cuenta el valor mínimo capaz de ser detectado. Por ejemplo, con 1000 permutaciones, el menor valor de p-values posible es  pvalue=1+11000+1=0.001998 .

A pesar de ello, la estabilidad y el error de la estimación varían mucho en función de los datos disponibles. Suele encontrarse como recomendación empírica emplear al menos 2000 iteraciones.

¿Cómo interpretar el p-value obtenido?

El p-value es la probabilidad de obtener un valor del estadístico igual o más extremo que el valor observado, si la hipótesis nula es cierta  P(datos|H0) . Es importante no confundir esto con la probabilidad de que la hipótesis nula sea cierta  P(H0|datos) . Si el valor es muy pequeño, se puede interpretar como una evidencia en contra de la hipótesis nula, pero no como una confirmación de las hipótesis alternativas.

## Regresión lineal bootstrap

In [186]:
import pandas as  pd

In [190]:
Data = pd.read_csv('House_Price_Regression.csv')

Data = Data.loc[:, ['latitude', 'longitude', 'price', 'size_in_m_2','no_of_bedrooms', 'no_of_bathrooms', 'quality_recode']]

Data.head()

latitude  longitude    price  size_in_m_2  no_of_bedrooms  \
0  25.113208  55.138932  2700000   100.242337               1   
1  25.106809  55.151201  2850000   146.972546               2   
2  25.063302  55.137728  1150000   181.253753               3   
3  25.227295  55.341761  2850000   187.664060               2   
4  25.114275  55.139764  1729200    47.101821               0   

   no_of_bathrooms  quality_recode  
0                2             2.0  
1                2             2.0  
2                5             2.0  
3                3             1.0  
4                1             2.0

In [206]:
Data['quality_recode'] = Data['quality_recode'].astype('category')

In [188]:
import sklearn

from sklearn.linear_model import LinearRegression

In [207]:
X = Data[['size_in_m_2', 'longitude', 'latitude', 'no_of_bedrooms', 'no_of_bathrooms', 'quality_recode']]
Y = Data['price']

In [202]:
X.head()

size_in_m_2  longitude   latitude  no_of_bedrooms  no_of_bathrooms  \
0   100.242337  55.138932  25.113208               1                2   
1   146.972546  55.151201  25.106809               2                2   
2   181.253753  55.137728  25.063302               3                5   
3   187.664060  55.341761  25.227295               2                3   
4    47.101821  55.139764  25.114275               0                1   

  quality_recode  
0            2.0  
1            2.0  
2            2.0  
3            1.0  
4            2.0

In [193]:
Y.head()

0    2700000
1    2850000
2    1150000
3    2850000
4    1729200
Name: price, dtype: int64

In [199]:
def varcharProcessing(X, varchar_process = "dummy_dropfirst"):
    
    dtypes = X.dtypes

    if varchar_process == "drop":   
        X = X.drop(columns = dtypes[dtypes == np.object].index.tolist())

    elif varchar_process == "dummy":
        X = pd.get_dummies(X,drop_first=False)

    elif varchar_process == "dummy_dropfirst":
        X = pd.get_dummies(X,drop_first=True)

    else: 
        X = pd.get_dummies(X,drop_first=True)
    
    X["intercept"] = 1
    cols = X.columns.tolist()
    cols = cols[-1:] + cols[:-1]
    X = X[cols]
    
    return X

In [208]:
X = varcharProcessing(X, varchar_process = "dummy_dropfirst")

X.head()

intercept  size_in_m_2  longitude   latitude  no_of_bedrooms  \
0          1   100.242337  55.138932  25.113208               1   
1          1   146.972546  55.151201  25.106809               2   
2          1   181.253753  55.137728  25.063302               3   
3          1   187.664060  55.341761  25.227295               2   
4          1    47.101821  55.139764  25.114275               0   

   no_of_bathrooms  quality_recode_1.0  quality_recode_2.0  quality_recode_3.0  
0                2                   0                   1                   0  
1                2                   0                   1                   0  
2                5                   0                   1                   0  
3                3                   1                   0                   0  
4                1                   0                   1                   0

In [209]:
model = LinearRegression().fit(X, Y)

In [210]:
residuals = np.sum( Y - model.predict(X) )

In [ ]:
residual